src/core/textrenderer/qgsfontmanager.cpp

```diff
@@ -0,0 +1,1122 @@
+  for ( const QString &replacement : replacements )
+  {
+    const thread_local QRegularExpression rxReplacement( QStringLiteral( "(.*?):(.*)" ) );
+    const QRegularExpressionMatch match = rxReplacement.match( replacement );
+    if ( match.hasMatch() )
+    {
+      mFamilyReplacements.insert( match.captured( 1 ), match.captured( 2 ) );
+      mLowerCaseFamilyReplacements.insert( match.captured( 1 ).toLower(), match.captured( 2 ) );
+    }
+  }
+}
+
+QMap<QString, QString> QgsFontManager::fontFamilyReplacements() const
+{
+  QgsReadWriteLocker locker( mReplacementLock, QgsReadWriteLocker::Read );
+  return mFamilyReplacements;
+}
+
+void QgsFontManager::addFontFamilyReplacement( const QString &original, const QString &replacement )
+{
+  QgsReadWriteLocker locker( mReplacementLock, QgsReadWriteLocker::Write );
+  if ( !replacement.isEmpty() )
+  {
+    mFamilyReplacements.insert( original, replacement );
+    mLowerCaseFamilyReplacements.insert( original.toLower(), replacement );
+  }
+  else
+  {
+    mFamilyReplacements.remove( original );
+    mLowerCaseFamilyReplacements.remove( original.toLower() );
+  }
+  storeFamilyReplacements();
+}
+
+void QgsFontManager::setFontFamilyReplacements( const QMap<QString, QString> &replacements )
+{
+  QgsReadWriteLocker locker( mReplacementLock, QgsReadWriteLocker::Write );
+  mFamilyReplacements = replacements;
+  mLowerCaseFamilyReplacements.clear();
+  for ( auto it = mFamilyReplacements.constBegin(); it != mFamilyReplacements.constEnd(); ++it )
+    mLowerCaseFamilyReplacements.insert( it.key().toLower(), it.value() );
+
+  storeFamilyReplacements();
+}
+
+QString QgsFontManager::processFontFamilyName( const QString &name ) const
+{
+  QgsReadWriteLocker locker( mReplacementLock, QgsReadWriteLocker::Read );
+  auto it = mLowerCaseFamilyReplacements.constFind( name.toLower() );
+  if ( it != mLowerCaseFamilyReplacements.constEnd() )
+    return it.value();
+  else
+    return name;
+}
+
+void QgsFontManager::storeFamilyReplacements()
+{
+  QStringList replacements;
+  for ( auto it = mFamilyReplacements.constBegin(); it != mFamilyReplacements.constEnd(); ++it )
+    replacements << QStringLiteral( "%1:%2" ).arg( it.key(), it.value() );
+  settingsFontFamilyReplacements.setValue( replacements );
+}
+
+void QgsFontManager::installUserFonts()
+{
+  QgsReadWriteLocker locker( mReplacementLock, QgsReadWriteLocker::Write );
+  const QString userProfileFontsDir = QgsApplication::qgisSettingsDirPath() + "fonts";
+  QStringList fontDirs { userProfileFontsDir };
+
+  fontDirs.append( mUserFontDirectories );
+
+  for ( const QString &dir : std::as_const( fontDirs ) )
+  {
+    const QDir localDir;
+    if ( !localDir.mkpath( dir ) )
+    {
+      QgsDebugMsg( QStringLiteral( "Cannot create local fonts dir: %1" ).arg( dir ) );
+      return;
+    }
+
+    const QFileInfoList fileInfoList = QDir( dir ).entryInfoList( QStringList( QStringLiteral( "*" ) ), QDir::Files );
```

Comment:
Maybe restrict the file list to *.ttf and *.otf just to play it safe?

Reply:
There's a situation we need to handle where the files won't have an extension -- when a font is downloaded direct from a url, we may not know the file name and extension. In this case we save the downloaded content just using the family name and no extension (qt doesn't provide any way to determine if it's say TTF or otf). So here we need to handle extension less files too.